### PR TITLE
feat: add AppModuleBasic for tendermint client

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

ref: #2009 

### Commit Message / Changelog Entry

```bash
feat: add `AppModuleBasic` for the 07-tendermint client and remove tendermint type registration from core IBC. Chains must register the `AppModuleBasic` of light clients.
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/types/codec.go [hunk 1]
```diff
@@ -17,6 +16,5 @@ func RegisterInterfaces(registry codectypes.InterfaceRegistry) {
 	connectiontypes.RegisterInterfaces(registry)
 	channeltypes.RegisterInterfaces(registry)
 	solomachine.RegisterInterfaces(registry)
-	ibctm.RegisterInterfaces(registry)
```
Comment (colin-axner) [inline modules/core/types/codec.go:20]: core IBC will no longer include clients by default. This is a feature in my opinion, as it makes core IBC neutral to which light clients should be registered. This is aligned with the IBC app registration approach which requires chain developers to add code to register an application. Luckily adding a client is extremely trivial

File: modules/light-clients/07-tendermint/module.go [hunk 2]
```diff
@@ -1,6 +1,53 @@
 package tendermint
```
Comment (colin-axner) [inline modules/light-clients/07-tendermint/module.go:1]: Was unsure if we wanted to make a separate package for the `module.go`? This file layout doesn't align with SDK as we removed the `types` directory as used by [traditional sdk modules](https://github.com/cosmos/cosmos-sdk/tree/main/x/gov), but newer sdk modules have a different [layout](https://github.com/cosmos/cosmos-sdk/tree/main/x/group). Now is probably the last chance to change things around without too much disruption. I don't have strong preferences

File: modules/light-clients/07-tendermint/module.go [hunk 3]
```diff
@@ -1,6 +1,53 @@
...
 
-// Name returns the IBC client name
-func Name() string {
```
Comment (colin-axner) [inline modules/light-clients/07-tendermint/module.go:4]: removed, seemed unused

File: modules/light-clients/07-tendermint/module.go (hunk 2 above)
Comment (charleenfei) [reply modules/light-clients/07-tendermint/module.go:1]: i guess i dont see the concrete benefit of making a separate package for module.go, and i dont see it being involved in any circular imports, so i don't see why it needs to be in a separate package. i think whichever we we choose to go we should just be consistent.

**Review by charleenfei — APPROVED** [review]

File: modules/light-clients/07-tendermint/module.go [hunk 4]
```diff
@@ -1,6 +1,53 @@
...
+)
+
+var _ module.AppModuleBasic = AppModuleBasic{}
+
+// AppModuleBasic defines the basic application module used by the tendermint light client.
+type AppModuleBasic struct{}
+
+// Name returns the tendermint module name.
+func (AppModuleBasic) Name() string {
 	return SubModuleName
```
Comment (crodriguezvega) [inline modules/light-clients/07-tendermint/module.go:21]: Not related to this PR, so could be done in a separate one, but I was just thinking if we should rename this `SubModuleName` to `ModuleName`? And `SubModuleName` is currently defined [errors.go](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint/errors.go#L8), but maybe could also moved to a new keys.go file like we have in other modules.

Same thing would apply to the solomachine light client, I guess.

File: modules/light-clients/07-tendermint/module.go (hunk 2 above)
Comment (colin-axner) [reply modules/light-clients/07-tendermint/module.go:1]: Agreed. I think for now it makes sense to just add the `module.go` file. We can always update in the future if we see a need to

File: modules/light-clients/07-tendermint/module.go (hunk 4 above)
Comment (colin-axner) [reply modules/light-clients/07-tendermint/module.go:21]: Yes agreed. Can you open an issue?

Comment (crodriguezvega) [reply modules/light-clients/07-tendermint/module.go:21]: Cool. I opened https://github.com/cosmos/ibc-go/issues/2833 and https://github.com/cosmos/ibc-go/issues/2834.

File: docs/migrations/v6-to-v7.md [hunk 5]
```diff
@@ -13,7 +13,31 @@ There are four sections based on the four potential user groups of this document
...
+### Light client registration
+
+Chains must explicitly register the types of any light client modules it wishes to integrate. 
+
+#### Tendermint registration
+
+To register the tendermint client, modify the `app.go` file to include the tendermint `AppModuleBasic`:
+
+```diff
+import (
```
Comment (crodriguezvega) [inline docs/migrations/v6-to-v7.md:25]: Could we fix the indentation to 4 spaces instead of 8?

File: docs/migrations/v6-to-v7.md [hunk 6]
```diff
@@ -13,7 +13,31 @@ There are four sections based on the four potential user groups of this document
...
+        ...
++       ibctm "github.com/cosmos/ibc-go/v6/modules/light-clients/07-tendermint"
+        ...
+)
+...
+
+ModuleBasics = module.NewBasicManager(
+                ...
```
Comment (crodriguezvega) [inline docs/migrations/v6-to-v7.md:33]: And here as well to 4 spaces?

**Review by crodriguezvega — APPROVED** [review]

File: modules/light-clients/07-tendermint/module.go [hunk 7]
```diff
@@ -1,6 +1,53 @@
...
 }
+
+// RegisterLegacyAminoCodec does nothing. The Tendermint client does not support amino.
+func (AppModuleBasic) RegisterLegacyAminoCodec(*codec.LegacyAmino) {}
+
+// RegisterInterfaces registers module concrete types into protobuf Any.
+func (AppModuleBasic) RegisterInterfaces(registry codectypes.InterfaceRegistry) {
+	RegisterInterfaces(registry)
+}
```
Comment (chatton) [inline modules/light-clients/07-tendermint/module.go:30]: same concerns as [over here](https://github.com/cosmos/ibc-go/pull/2826#discussion_r1034523499).

File: testing/simapp/app.go [hunk 8]
```diff
@@ -153,6 +154,7 @@ var (
 		crisis.AppModuleBasic{},
 		slashing.AppModuleBasic{},
 		ibc.AppModuleBasic{},
+		ibctm.AppModuleBasic{},
```
Comment (chatton) [inline testing/simapp/app.go:157]: is it worth adding a comment here to explicitly state that we are providing a module for the tendermint light client so the interfaces are registered? Maybe overkill, will leave it to you!

**Review by chatton — APPROVED** [review]
Looks great, just left a few comments

File: testing/simapp/app.go (hunk 8 above)
Comment (colin-axner) [reply testing/simapp/app.go:157]: Going to leave it for now. I think we can revisit in https://github.com/cosmos/ibc-go/issues/2008

**Comment by colin-axner** [comment]
tagging automerge, please open an issue or pr if there any further requested changes

**Pull request merged**